### PR TITLE
Prevent private functions from being overridden

### LINKED ISSUE #8417: Private method collision in hierarchy
## Description

It seems a private method could be overridden!?

## Environment

- Compiler version: 0.5.16

## Steps to Reproduce

```solidity
pragma solidity ^0.5.0;

contract A {
    function v() public view returns(uint256) { return f(); }
    function f() private pure returns(uint256) { return 1; }
}

contract B is A {
    function v() public view returns(uint256) { return f() + super.v(); }
    function f() private pure returns(uint256) { return 2; }
}
```
=>
```
(new B()).v() == 4
```

### PR DESCRIPTION
**Pull request by Marenz**
fixes #8417

### CODE REVIEW

**Comment by Marenz** [comment]
the failed zepplin test seems to be because it uses `override` etc which this compiler version doesn't support. Not sure why it does that.. maybe the test files are fetched externally.

Don't know the problem with the solcjs test. Help appreciated.

**Comment by leonardoalt** [comment]
@axic any idea what the solc-js problem could be here? Can it be related to the release version?

**Comment by ekpyron** [comment]
We probably need to patch soljson.js here, let me check.

**Comment by ekpyron** [comment]
So adding
```
# Patch soljson.js to provide backwards-compatibility with older emscripten versions
echo ";/* backwards compatibility */ Module['Runtime'] = Module;" >> $BUILD_DIR/libsolc/soljson.js
```
to ``build_emscripten.sh`` right before copying to the ``upload/`` directory should fix this, but thinking about it, it shouldn't be necessary and may be an issue in solc-js - I'll quickly check there.

**Comment by ekpyron** [comment]
Hm, that's weird... the ``soljson.js`` built by CI here, *does* correctly contain that workaround...

**Comment by ekpyron** [comment]
Ah, right, I know why this happens. This is (IMHO) pretty annoying about the ``solc-js`` tests... it will download the most recent version of soljson.js on its own (i.e. 0.6.4) and run against that as well without being asked... and in that we don't have the workarounds anymore...

**Comment by ekpyron** [comment]
I.e. the failure is entirely harmless, but it will be hard to fix...

**Comment by ekpyron** [comment]
One option would be to add
``sed -i -e 's/latest/v0.5.16+commit.9c3226ce/' test/package.js``
Right before running the tests in ``test/solcjsTests.sh``.
Or actually maybe better ``sed -i -e 's/latest/v0.5.0+commit.1d4f565a/' test/package.js``.

It's running on solc-js version 0.5.0 and we don't want to test that against latest (i.e. 0.6.4), since that is bound to break. Actually we don't want to test against any remote version here at all :-D... but the sanest choice, since we kind of have to, would be ``v0.5.0``.

**Comment by ekpyron** [comment]
The other test run: ``t_ems_compile_ext_zeppelin``: if we want that to pass we need to downgrade to a version of zeppelin that is 0.5 compatible. I.e. replace ``master`` in
``test/externalTests/zeppelin.sh`` with some appropriate branch/tag/commit.

**Comment by leonardoalt** [comment]
Sigh.. would be fine with not passing the zeppelin test

File: Changelog.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+### 0.5.17 (unreleased)
+
+Bugfixes:
+ * Type Checker: Prevent overriding of private functions.
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
 * Type Checker: Disallow overriding of private functions.
```

**Comment by Marenz** [comment]
Update:
* Added test for multiple inheritance
* Using version 2.5.0 of zepplin, which is before they updated to 0.6.0
* Added fix daniel suggested to solcjs test

File: libsolidity/analysis/ContractLevelChecker.cpp [hunk 2]
```diff
@@ -203,6 +203,9 @@ void ContractLevelChecker::checkFunctionOverride(FunctionDefinition const& _func
 			stateMutabilityToString(_function.stateMutability()) +
 			"\"."
 		);
+	if (_function.visibility() == Declaration::Visibility::Private)
```
Comment (chriseth) [inline libsolidity/analysis/ContractLevelChecker.cpp]: This actually checks "private functions cannot override". Can you add a test for different combinations of publc / private in the base an derived?

Comment (Marenz) [reply libsolidity/analysis/ContractLevelChecker.cpp]: Done. All fail as they should :)

Comment (chriseth) [reply libsolidity/analysis/ContractLevelChecker.cpp]: But the error messages are wrong: https://github.com/ethereum/solidity/pull/8455/files#diff-302ae33fa6301f5b2401b6e8cd02484a

You need to test for `_super.visibility()`

Comment (Marenz) [reply libsolidity/analysis/ContractLevelChecker.cpp]: oh, true. Updated

**Comment by chriseth** [comment]
Looks good! I'll make this the release PR directly.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**